plasmapy/formulary/collisions/tools/collisional_analysis.py

```diff
@@ -0,0 +1,337 @@
+
+
+    Examples
+    --------
+
+
+    """
+
+    # Validate number of ions
+    if len(ions) != 2:
+        raise ValueError(
+            "Argument 'ions' can only take two input values. Instead "
+            f"received {len(ions)} input values."
+        )
+
+    # Validate n_step argument
+    if not isinstance(n_step, int):
+        raise TypeError(
+            "Argument 'n_step' is of incorrect type, type of "
+            f"{type(n_step)} received instead. While 'n_step' must be "
+            "of type int."
+        )
+
+    # Validate scaling arguments
+    for arg in (density_scale, velocity_scale, temperature_scale):
+        if not isinstance(arg, (int, float, numbers.Integral)):
+            raise TypeError(
+                "Scaling argument is of incorrect type, type of "
+                f"{type(arg)} received instead. Scaling argument "
+                "should be of type float or int."
+            )
+
+    # Define differential equation function
+    def df_eq(
+        r_0,
+        r_n,
+        n_1,
+        n_2,
+        v_1,
+        T_1,
+        T_2,
+        ions,
+        n_step,
+        density,
+        velocity,
+        temperature,
+    ):
+        # Initialize the alpha-proton charge and mass ratios.
+        z_1 = ions[0].charge_number
+        mu_1 = ions[0].mass_number
+
+        z_2 = ions[1].charge_number
+        mu_2 = ions[1].mass_number
+
+        # Initialise.
+        d_r = (r_0 - r_n) / (1.0 * n_step)
+        d_r = d_r.value
+
+        # Define constant
+        k = 0.174
+
+        # Define Coulomb log for mixed ion collisions
+        def lambda_ba(
+            theta,
+            n_1,
+            n_2,
+            z_1,
+            z_2,
+            mu_1,
+            mu_2,
+        ):
+            return 9 + np.log(
+                np.sqrt(n_2 * z_2**2 / theta * n_1 * z_1**2 + 1)
+                * ((z_1 * z_2 * (mu_1 + mu_2)) / (theta + (mu_2 / mu_1)))
+                * ((theta + mu_2 / mu_1) / (z_1 * z_2 * (mu_1 + mu_2)))
+            )
+
+        # Loop.
+        for i in range(n_step):
+            r = r_n.value + ((i + 1) * d_r)
+
+            eta = n_2 / n_1
+            theta = T_2 / T_1
+
+            l_ba = lambda_ba(theta, n_1, n_2, z_1, z_2, mu_1, mu_2)
+
+            n_1 = n_1 * (r / r_n) ** density
+            v_1 = v_1 * (r / r_n) ** velocity
+            T_1 = T_1 * (r / r_n) ** temperature
+
+            d_theta = (
+                d_r
+                * l_ba
+                * k
+                * (np.sqrt(mu_1 * mu_2 * z_1 * z_2) * (1 - theta) * (1 + eta * theta))
+                / (np.sqrt(mu_1 / mu_2 + theta) ** 3)
+            )
+
+            print(theta, d_theta)
+            theta = theta + d_theta
+
+        return theta
+
+    vars = [r_0, r_n, n_1, n_2, v_1, T_1, T_2]
+
+    d_type = []
+    for var in vars:
+        if hasattr(var, "__len__"):
+            d_type.append(True)
+        else:
+            d_type.append(False)
```

Comment:
I'm wondering if we could do this via a comprehension with an `if` clause in there. 🤔